crates/bevy_asset/src/loader.rs

```diff
@@ -1,274 +1,397 @@
     }
-}
-
-/// The result of loading an asset of type `T`.
-#[derive(Debug)]
-pub struct AssetResult<T> {
-    /// The asset itself.
-    pub asset: Box<T>,
-    /// The unique id of the asset.
-    pub id: HandleId,
-    /// Change version.
-    pub version: usize,
-}
 
-/// An event channel used by asset server to update the asset storage of a `T` asset.
-#[derive(Debug)]
-pub struct AssetLifecycleChannel<T> {
-    /// The sender endpoint of the channel.
-    pub sender: Sender<AssetLifecycleEvent<T>>,
-    /// The receiver endpoint of the channel.
-    pub receiver: Receiver<AssetLifecycleEvent<T>>,
-}
+    /// Gets the source asset path for this load context.
+    pub fn asset_path(&self) -> &AssetPath {
+        &self.asset_path
+    }
 
-/// Events for the [`AssetLifecycleChannel`].
-pub enum AssetLifecycleEvent<T> {
-    /// An asset was created.
-    Create(AssetResult<T>),
-    /// An asset was freed.
-    Free(HandleId),
-}
+    /// Gets the source asset path for this load context.
+    pub async fn read_asset_bytes<'b>(
+        &mut self,
+        path: &'b Path,
+    ) -> Result<Vec<u8>, AssetReaderError> {
+        let mut reader = self.asset_server.reader().read(path).await?;
+        let mut bytes = Vec::new();
+        reader.read_to_end(&mut bytes).await?;
+        todo!("populate hash here");
+        // self.loader_dependencies
+        //     .insert(AssetPath::new(path.to_owned(), None));
+        Ok(bytes)
+    }
 
-/// A trait for sending lifecycle notifications from assets in the asset server.
-pub trait AssetLifecycle: Downcast + Send + Sync + 'static {
-    /// Notifies the asset server that a new asset was created.
-    fn create_asset(&self, id: HandleId, asset: Box<dyn AssetDynamic>, version: usize);
-    /// Notifies the asset server that an asset was freed.
-    fn free_asset(&self, id: HandleId);
-}
-impl_downcast!(AssetLifecycle);
-
-impl<T: AssetDynamic> AssetLifecycle for AssetLifecycleChannel<T> {
-    fn create_asset(&self, id: HandleId, asset: Box<dyn AssetDynamic>, version: usize) {
-        if let Ok(asset) = asset.downcast::<T>() {
-            self.sender
-                .send(AssetLifecycleEvent::Create(AssetResult {
-                    asset,
-                    id,
-                    version,
-                }))
-                .unwrap();
+    /// Retrieves a handle for the asset at the given path and adds that path as a dependency of the asset.
+    /// If the current context is a normal [`AssetServer::load`], an actual asset load will be kicked off immediately, which ensures the load happens
+    /// as soon as possible.
+    /// If the current context is an [`AssetServer::load_direct_async`] (such as in the [`AssetProcessor`](crate::processor::AssetProcessor)),
+    /// a load will not be kicked off automatically. It is then the calling context's responsibility to begin a load if necessary.
+    pub fn load<'b, A: Asset>(&mut self, path: impl Into<AssetPath<'b>>) -> Handle<A> {
```

Comment:
Hmm, I'd like to pass the `AssetLoader::Settings` here, but `A: Asset` isn't `AssetLoader`. I feel like maybe there should be a trait for `LoadableAsset` so that it's possible to pass `<A as LoadableAsset>::Loader::Settings` here.

Reply:
If A doesn't have a loader this will print a "missing loader" error to the console and the asset LoadState will enter the Failed state. The handle will remain active until it is dropped, but it won't have access to an asset.